Feature/48 답변 작성 UI 개발

yoonchanchoi opened the pull request:
#### close #48

### ✏️ 개요

 답변 작성 UI 개발

### 💻 작업 사항

 답변 작성 UI 개발

### 📄 리뷰 노트

텍스트 뷰의  색 변경에 대해서 databidning으로 처리해보고 싶었는데 잘 안되서 어쩔 수 없이 코드에서 setcolor로 바꾸었습니다. 이 부분 더 좋은 방법이 있다면 말씀해주세용^^


### 📱결과 화면(optional)
<img src="https://github.com/abloom-AOS/MERY_Android/assets/74814647/ed377ff5-e4fa-4b62-baa4-bec61a6468c1" width="200" height="400">
<img src="https://github.com/abloom-AOS/MERY_Android/assets/74814647/30f7714d-bacf-4521-9332-df894b3222e0" width="200" height="400">
<img src="https://github.com/abloom-AOS/MERY_Android/assets/74814647/26ba35c5-4d4e-4f25-b430-1af952e02e4f" width="200" height="400">
<img src="https://github.com/abloom-AOS/MERY_Android/assets/74814647/619d07b1-d3e6-4e3c-8890-83100e585f1c" width="200" height="400">
<img src="https://github.com/abloom-AOS/MERY_Android/assets/74814647/dcfdb351-1db6-4924-bd88-543c2d699db0" width="200" height="400">

yoonchanchoi committed:
[Feat/#36] write_answer 수정

yoonchanchoi committed:
Merge branch 'develop' into feature/48-답변-작성-UI-개발

yoonchanchoi committed:
[Feat/#36] write_answer xml 수정

yoonchanchoi committed:
[Feat/#36] write_answer 뷰 구현 및 databinding color값 수정

yoonchanchoi committed:
[Feat/#36] write_answer 필요없는 주석 및 코드 제거

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment Color값 수정

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment dialogn 버튼 text 수정

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment dialogn 나가기 버튼시 화면 이동 위치 수정

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment xml 파일 hint Stirng파일 추가

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment 텍스트길이 변수 명칭 수정

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment 코드 정렬

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment xml 이름 변경

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment 미사용 코드(arg) 삭제

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment 주석 삭제

yoonchanchoi committed:
Merge branch 'develop' into feature/48-답변-작성-UI-개발

yoonchanchoi committed:
[Feat/#36] WriteAnswerFragment popback 변경

ki960213 commented on app/src/main/java/com/abloom/mery/presentation/ui/writeanswer/WriteAnswerFragment.kt:
```diff
@@ -15,15 +19,97 @@ import dagger.hilt.android.AndroidEntryPoint
+        repeatOnStarted {
+            writeAnswerViewModel.question.collect {
+                binding.question = it?.content ?: QUESTION_NULL
+            }
```
널일 때 무시하고 싶다면 `filterNotNull()`로 널 처리를 해도 될 것 같아요. 😄

yoonchanchoi replied:
아하 감사합니다. 늘 더 좋은 방법을 찾으시다니.
솔직히 몰랐습니다 저건ㅋㅋ

ki960213 committed:
[Design/#48] 빨강 색상 추가

ki960213 committed:
[Feat/#48] xml에서 Int 타입을 TextView의 text로 설정할 수 있도록 바인딩 어댑터 구현

ki960213 committed:
[Feat/#48] 피그마대로 문자열 리소스 변경

ki960213 committed:
[Feat/#48] WriteAnswerViewModel에 답변의 최대 길이 상수 추가

ki960213 committed:
[Feat/#48] fragment_write_answer.xml에서 데이터 바인딩을 사용하여 기능 구현

ki960213 committed:
[Feat/#48] 답변 작성 화면 기능 구현

- 뒤로 갈 때 답변이 비어있지 않으면 화면을 나갈지 묻는 다이얼로그 보여주기
- 앱바의 내비게이션 아이콘을 클릭하면 뒤로 가기 실행
- 소프트 키보드 변화에 맞춰 화면 크기 변경

ki960213 committed:
[Style/#48] 코드 정렬

ki960213 commented on app/src/main/java/com/abloom/mery/presentation/ui/writeanswer/WriteAnswerFragment.kt:
```diff
@@ -15,15 +19,83 @@ import dagger.hilt.android.AndroidEntryPoint
+        setupDataBinding()
+    }
+
+    private fun setupWindowInsetsListener(view: View) {
+        ViewCompat.setOnApplyWindowInsetsListener(view) { _, insets ->
+            val imeVisible = insets.isVisible(WindowInsetsCompat.Type.ime())
+            val imeHeight = insets.getInsets(WindowInsetsCompat.Type.ime()).bottom
+            if (imeVisible) {
+                binding.root.setPadding(0, 0, 0, imeHeight)
+            } else {
+                binding.root.setPadding(0, 0, 0, 0)
+            }
+            insets
         }
```
소프트 키보드가 보여지는 것을 감지해 화면 크기를 변경하도록 했습니다.

yoonchanchoi replied:
매니페스트에서 키보드가 띄워주는 위치를 변경할 생각은 해봤는데 그렇게되면 저희가 단일Activity라 적용하지 못했는데 저런 방법도 있군요.b!!!! 배웠습니다. 생각의 전환이네용 ㅎㅎ

ki960213 replied:
공식 문서만 봐도 해결할 수 있는 문제가 많더라구요 😄 
https://developer.android.com/develop/ui/views/layout/sw-keyboard?hl=ko

yoonchanchoi replied:
한번 봐야겠습니당^^

ki960213 commented on app/src/main/java/com/abloom/mery/presentation/ui/writeanswer/WriteAnswerFragment.kt:
```diff
@@ -15,15 +19,83 @@ import dagger.hilt.android.AndroidEntryPoint
+    }
+
+    private fun setupAppBar() {
+        binding.appbarWriteAnswer.setOnNavigationClick {
+            requireActivity().onBackPressedDispatcher.onBackPressed()
         }
+
```
뒤로 가기 버튼 클릭 시에도 다이얼로그가 보여지게 했습니다.

yoonchanchoi replied:
감사합니다. backpass도 생각했었는데 놓쳤네요..ㅠㅠ

ki960213 commented on app/src/main/java/com/abloom/mery/presentation/ui/writeanswer/WriteAnswerFragment.kt:
```diff
@@ -15,15 +19,83 @@ import dagger.hilt.android.AndroidEntryPoint
-    private val viewModel: WriteAnswerViewModel by viewModels()
+    private val writeAnswerViewModel: WriteAnswerViewModel by viewModels()
+
+    private val onBackPressedCallback = object : OnBackPressedCallback(true) {
+        override fun handleOnBackPressed() {
+            if (writeAnswerViewModel.answer.value.isNotBlank()) {
+                showBackConfirmDialog()
+            } else {
+                findNavController().popBackStack()
+            }
+        }
+    }
```
뒤로 가기 버튼을 클릭하거나 앱바의 내비게이션 버튼 클릭 시 사용자가 작성한 답변이 비어있지 않을 때만 화면을 나갈 건지 묻는 다이얼로그를 띄우도록 했습니다.

yoonchanchoi replied:
놓친 부분 감사합니다!

ki960213 committed:
[Ref/#48] 불필요한 리소스 제거

ki960213 commented on app/src/main/res/values/colors.xml:
```diff
@@ -2,6 +2,7 @@
 <resources>
     <color name="black">#FF000000</color>
     <color name="white">#FFFFFFFF</color>
+    <color name="red">#FFFF0000</color>
```
빨간색이 많이 쓰이기 때문에 색상을 추가했습니다.

yoonchanchoi replied:
진짜 추가해야되나 말아야되나 해서 일단 여쭈어보려고 했습니다. 감사합니다.

ki960213 commented on app/src/main/res/values/strings.xml:
```diff
@@ -118,4 +118,14 @@
     <string name="profilemenu_since_marriage_date_format">결혼한지 %d일 지났어요.</string>
     <string name="profilemenu_login_confirm_dialog_title">로그인이 필요한 서비스예요.</string>
 
+    <string name="writeanswer_limit_textsize_format">" / %d"</string>
```
답변의 최대 길이를 동적으로 표시할 수 있도록 했습니다.

ki960213 commented on app/src/main/res/layout/fragment_write_answer.xml:
```diff
@@ -5,24 +5,99 @@
+            android:layout_width="match_parent"
             android:layout_height="wrap_content"
             app:action="@string/all_complete"
+            app:isActionEnabled="@{!viewModel.answer.blank &amp;&amp; viewModel.answer.length &lt;= viewModel.ANSWER_LIMIT_SIZE}"
```
답변이 비어있지 않고 최대 길이보다 작거나 같을 때만 액션 버튼이 활성화되도록 했습니다.

yoonchanchoi replied:
&lt;=  이 <= 이거 군요
확실히 아직 제가 databinding 미숙한 가봅니다. 
저보다 훨씬 자연스럽게 쓰시네요..ㄷㄷ
감사합니다.

ki960213 replied:
데이터 바인딩을 쓸 때 가끔 특수문자를 이스케이프 해야 할 때가 있더라구요 ㅎㅎ

ki960213 commented on app/src/main/res/layout/fragment_write_answer.xml:
```diff
@@ -5,24 +5,99 @@
+                android:inputType="text|textMultiLine"
+                android:maxLength="500"
+                android:text="@={viewModel.answer}"
+                android:textColor="@{viewModel.answer.blank ? @color/neutral_40 : @color/neutral_50}"
```
답변이 비어있을 때 힌트의 색상과 답변의 색상이 다르기 때문에 바인딩 어댑터를 사용하여 동적으로 색상을 선택할 수 있도록 했습니다.

yoonchanchoi replied:
감사합니다.!!
피그마에서 놓친 부분이였습니다.
조심하겠습니다.

ki960213 commented on app/src/main/res/layout/fragment_write_answer.xml:
```diff
@@ -5,24 +5,99 @@
+                    style="@style/footnote"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="@{viewModel.answer.length}"
```
제가 TextView의 바인딩 어댑터를 구현했기 때문에, Int 타입도 바인딩 어댑터를 사용할 수 있습니다.

yoonchanchoi replied:
진짜 고생 많으셨습니다. 바인딩 어댑터는 찾아보니깐 일일이 해당 경우의 수를 구현하는 거던데 감사합니다.

ki960213 commented on app/src/main/res/layout/fragment_write_answer.xml:
```diff
@@ -5,24 +5,99 @@
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="@{viewModel.answer.length}"
+                    android:textColor="@{viewModel.answer.length >= viewModel.ANSWER_LIMIT_SIZE ? @color/red : @color/neutral_40}" />
```
바인딩 어댑터를 사용하여 답변의 길이에 따라 텍스트 색상을 변하도록 했습니다.

yoonchanchoi replied:
감사합니다.
뷰모델에서 150으로 limit를 정해서 하니깐 뷰 쪽 코드가 1개라도 줄어드는 군요.ㅎㅎ

ki960213 replied:
`viewModel.ANSWER_LIMIT_SIZE` 대신 그냥 150으로 하드코딩하는 게 코드는 더 짧은 것 같아요 😓 
다만 중복을 없애서 유지보수성이 향상되었다고 생각해요.

yoonchanchoi replied:
맞습니다. 저도 동감합니당^^

ki960213 commented on app/src/main/res/layout/fragment_write_answer.xml:
```diff
@@ -5,24 +5,99 @@
+                    style="@style/footnote"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="@{@string/writeanswer_limit_textsize_format(viewModel.ANSWER_LIMIT_SIZE)}"
```
문자열 리소스 포맷 기능을 사용하여 유지보수성을 향상했습니다.

yoonchanchoi replied:
뭔가 코드 가독성을 위해서 어디에서 변수를 선언하고 처리할지도 많이 중요하네용
fragment코드도 엄청 많이 줄어들었네용. 감사합니다.!!
보고 배웁니다. 그리고 확실히 ㅋㅋㅋ좋네용 좀 더 익숙해져야겠습니당^^

ki960213 reviewed (approved):
제가 데이터 바인딩과 바인딩 어댑터를 사용하도록 코드를 변경했고, 아래처럼 디자인을 약간 수정했습니다.

|소프트 키보드가 숨겨졌을 때|소프트 키보드가 보여질 때|답변 작성 확인 다이얼로그의 메세지|
|--|--|--|
|![image](https://github.com/abloom-AOS/MERY_Android/assets/123928686/874fa82d-10b1-4a2f-a9b9-99dcfde5c142)|![image](https://github.com/abloom-AOS/MERY_Android/assets/123928686/79fa14a9-6564-4402-bfe6-e7b63b11ed36)|![image](https://github.com/abloom-AOS/MERY_Android/assets/123928686/56977b71-5e42-4b27-89df-347c6af86e9a)|

그리고 사용성을 고려해 일부 로직 변경했습니다. 자세한 내용은 제 comment를 참고해주세요 😄

leeugun123 commented on app/src/main/java/com/abloom/mery/presentation/ui/signup/BrideGroomSelectionFragment.kt:
```diff
@@ -41,7 +41,12 @@ class BrideGroomSelectionFragment :
+                android.R.anim.fade_out,
+                android.R.anim.fade_in,
+                android.R.anim.fade_out
+            )
```
이 부분 제가 이미 변경해서 머지 하였습니다.

yoonchanchoi replied:
감사합니다!!

ki960213 committed:
[Fix/#48] 답변 최대 길이보다 길 때 글자 수가 빨간색으로 보이도록 수정

leeugun123 reviewed (approved):
구현하시느라 고생많으셨습니다.!

ki960213 committed:
[Fix/#48] 소프트 키보드가 보여질 때 내비게이터 바의 높이에 영향 받지 않도록 수정
